### DISCUSSION TITLE
Audio Distortion After About One Second

### DISCUSSION THREAD

**Discussion by frohro** [pschatzmann/Adafruit_TinyUSB_Arduino#7] [Q&A] [answered]
@pschatzmann , I am experiencing a problem with audio distortion occurring after about one second of transmitting data.  My code is below:
```
#include <cmath>
#include "Adafruit_TinyUSB.h"

const float sampleRate = 48000.0;
const float frequencyLeft = 400;
const float frequencyRight = 400;
const float amplitude = 10000.0;
const float twoPi = 2.0 * M_PI;

Adafruit_USBD_Audio usb;

size_t readCB(uint8_t* data, size_t len, Adafruit_USBD_Audio& ref) {
  int16_t* data16 = (int16_t*)data;
  Serial.println(len);
  size_t samples = len / sizeof(int16_t);
  size_t result = 0;
  static float phaseLeft = 0.0;
  static float phaseRight = 0.0;
  float phaseIncrementLeft = twoPi * frequencyLeft / sampleRate;
  float phaseIncrementRight = twoPi * frequencyRight / sampleRate;

  for (size_t j = 0; j < samples; j += 2) {
    data16[j] = amplitude * sin(phaseLeft); // Left channel
    data16[j + 1] = amplitude * sin(phaseRight); // Right channel
    phaseLeft += phaseIncrementLeft;
    phaseRight += phaseIncrementRight;
    if (phaseLeft >= twoPi) phaseLeft -= twoPi;
    if (phaseRight >= twoPi) phaseRight -= twoPi;
    result += sizeof(int16_t) * 2;
  }
  return result;
}

void setup() {
  Serial.begin(115200);

  // Start USB device as Microphone
  usb.setReadCallback(readCB);
  usb.begin(sampleRate, 2, 16); 

  if (TinyUSBDevice.mounted()) {
    TinyUSBDevice.detach();
    delay(10);
    TinyUSBDevice.attach();
  }
}

void loop() {
  // optionally use LED to display status
  usb.updateLED();
}
```
I first saw this in quisk, an SDR software I use, set for audio frequency.  This is shown in the first video.  Then I verified it in Audacity.  This is shown in the second video below.  I have my Ubuntu box set to use the default pulse source.   I start out with my microphone (the little squiggle).  Then I go to the Pico sending a sinusoid over USB by plugging it in the USB port.  You cannot see the distortion looking at the wave in the time domain, but if you look at it in the frequency domain, you can.  The distortion does not appear immediately.  See the first spectrum and the good results.  But after about a second, it starts.  See the second spectrum.  Note I am making the spectrum so it starts exactly on one cycle and stops ready for the next cycle to avoid spectral leakage.  The distortion remains roughly constant after the first second.

https://github.com/user-attachments/assets/bc371b39-4ec3-4747-90a9-f78a8347c44d
https://github.com/user-attachments/assets/229240a3-e287-4255-85c9-2942e83dda93

I think this is a problem with the Audio part of the TinyUSB library, but it might be something I don't understand in my code.  Can anyone see a bug in my code?

**Comment by h5n1xp** [comment]
Hi @frohro you a doing a lot of processing (the M0 in the RP2040 has very weak float performance, you might get better results using a Sin() lookup table, also using fixed point maths is a trick I used to use back in my MC68000 days) in the USB callback, my suggestion is to render the audio to a separate buffer using the second core, and then use the callback to just copy the buffer to the USB stack.

I believe the readCB() function is called somewhat sporadically (whenever time is allocated on the bus) so the amount of time available to process it is unpredictable.

I can't get Serial and USB Audio functionality to work at the same time, so I have no idea how big the USB stack buffer is at the moment.

**Reply by frohro** [reply]
@h5n1xp, thanks for the suggestion.  I tried addressing it by just precomputing a buffer in the setup() function since sinewave are periodic, but the result is identical as far as I can tell.  Here is my code in case I made some obvious blunder.
```
#include <cmath>
#include "Adafruit_TinyUSB.h"
#include <vector>

const float sampleRate = 48000.0;
const float frequencyLeft = 400;
const float frequencyRight = 400;
const float amplitude = 10000.0;
const float twoPi = 2.0 * M_PI;

// Wavetable declarations
const size_t TABLE_SIZE = 1024;  // Power of 2 for efficient wrapping
std::vector<int16_t> sineTableLeft(TABLE_SIZE);
std::vector<int16_t> sineTableRight(TABLE_SIZE);
size_t tableIndex = 0;

Adafruit_USBD_Audio usb;

size_t readCB(uint8_t* data, size_t len, Adafruit_USBD_Audio& ref) {
  int16_t* data16 = (int16_t*)data;
  Serial.println(len);
  size_t samples = len / sizeof(int16_t);
  size_t result = 0;

  for (size_t j = 0; j < samples; j += 2) {
    data16[j] = sineTableLeft[tableIndex];
    data16[j + 1] = sineTableRight[tableIndex];
    tableIndex = (tableIndex + 1) % TABLE_SIZE;
    result += sizeof(int16_t) * 2;
  }
  return result;
}

void setup() {
  Serial.begin(115200);

  // Precompute sine tables
  float phaseIncrementLeft = twoPi * frequencyLeft / sampleRate;
  float phaseIncrementRight = twoPi * frequencyRight / sampleRate;

  for (size_t i = 0; i < TABLE_SIZE; i++) {
    float phase = twoPi * i / TABLE_SIZE;
    sineTableLeft[i] = amplitude * sin(phase);
    sineTableRight[i] = amplitude * sin(phase);
  }

  // Start USB device as Microphone
  usb.setReadCallback(readCB);
  usb.begin(sampleRate, 2, 16); 

  if (TinyUSBDevice.mounted()) {
    TinyUSBDevice.detach();
    delay(10);
    TinyUSBDevice.attach();
  }
}

void loop() {
  // optionally use LED to display status
  usb.updateLED();
}
```
I am wondering if it has to do with your finding that the sample rate is not constant.  I am doing this on Ubuntu 24.10.  I think I need to investigate how to make sure data is being received at my sampleRate.

**Reply by h5n1xp** [reply]
I'll give your code a test when I get some time.

But yes I'm am concerned about the weird sample rate reporting. My gut feeling is that the Pico isn't acknowledging the transfer in time, if at all! And I'm making an assumption that isochronous transfers require the device to acknowledge completion. I know almost nothing about low level USB!

**Reply by h5n1xp** [reply]
I've spent some time reading the USB 1.1 spec, and I think we need to be aware of the synchronisation field in the transfer packet... more reading to do...

**Comment by frohro** [comment]
Here is some more data on the problem.  It seems to be missing data every so often.  If you look at the first second when things are good, it looks like this with a 12 kHz sinusoid (48 kHz sampling, so this is four points per sinusoid).  
![image](https://github.com/user-attachments/assets/8ad5574b-f3e9-47ee-917f-56c29f2c7f68)
If you look later, when things are not working so well it looks like this if you go the a revealing zoom level. 
![image](https://github.com/user-attachments/assets/6283fde3-c616-45fa-95cc-b33b87262503)
Zoomed out a little more it looks like this:  
![image](https://github.com/user-attachments/assets/9146af6d-335b-4860-b556-092be16cd12f)
and a little more zoom:  
![image](https://github.com/user-attachments/assets/290581fd-fdc1-45c3-9fea-bf7c74d991c1)

Overclocking the Pico does not seem to help.  This is what it looks like at 312.6 MHz.  
![image](https://github.com/user-attachments/assets/0d79a1c5-917f-4f83-9967-d0ee0dbb4559)
This is what it looks like at 156.3 MHz:  
![image](https://github.com/user-attachments/assets/797f2f37-0ea6-4b16-84a8-0f605271ff77)
So changing from 125 MHz to 156.3 MHz doesn't help, and neither does changing to twice that frequency.  The frequency 156.3 was chosen because it is an integer multiple of 48000 Hz.
Changing the sample rate to 8 kHz and the sample frequency to 2 kHz does not help.  It looks like the problems happen about 1/6 as often, which indicates it is probably some periodic interrupt or something causing the problem.  Here is how that looks.
![image](https://github.com/user-attachments/assets/96296767-548d-4476-b1fc-da50bfa330f7)
You can see that the data drop does not always happen, just usually.

**Comment by pschatzmann** [comment]
Are you sure that the issue is not coming from your sine implementation ? 
Here is my result for your setup on a Pico W2 tested with Linux Mint on an old McBook Air:

![image](https://github.com/user-attachments/assets/bd6fe726-24be-441d-9e49-26e743af7add)

I was using my [AudioTools](https://github.com/pschatzmann/arduino-audio-tools) library to generate the sine:

```C++

#include "Adafruit_TinyUSB.h"
#include "AudioTools.h"

AudioInfo info(48000, 2, 16);
SineWaveGenerator<int16_t> sineWave(32000);                // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<int16_t> sound(sineWave);             // Stream generated from sine wave
Adafruit_USBD_Audio usb;

size_t readCB(uint8_t* data, size_t len, Adafruit_USBD_Audio& ref) {
  return sound.readBytes(data, len);
}

void setup() {
  Serial.begin(115200);

  // Setup sine wave
  sineWave.begin(info, 12000.0);

  // Start USB device as Audio Source
  usb.setReadCallback(readCB);
  usb.begin(info.sample_rate, info.channels, info.bits_per_sample);

  // If already enumerated, additional class driverr begin() e.g msc, hid, midi won't take effect until re-enumeration
  if (TinyUSBDevice.mounted()) {
    TinyUSBDevice.detach();
    delay(10);
    TinyUSBDevice.attach();
  }
}

void loop() {
  // use LED do display status
  usb.updateLED();
}

```

**Reply by pschatzmann** [reply, marked as answer]
I was testing with a Macbook Air running Mint Linux: Your test sketch with a __Pico 2 W__ gives the following for me
![image](https://github.com/user-attachments/assets/2d3521ba-8305-4613-b8d7-7a7eaa963419)

And I am getting the same with a regular __Pico__

![image](https://github.com/user-attachments/assets/61203174-9481-4f77-ab49-143505b58534)


So I can't reproduce your issue: Maybe it is related to your Desktop!

**Reply by frohro** [reply]
Thanks @pschatzmann for the help!  I tried to use your code above using your AudioTools, but I get a compile error:
```In file included from /home/frohro/.platformio/packages/framework-arduinopico/libraries/FreeRTOS/src/variantHooks.cpp:37:
/home/frohro/.platformio/packages/framework-arduinopico/libraries/FreeRTOS/src/variantHooks.cpp: In function 'void __usb(void*)':
.pio/libdeps/pico/Adafruit TinyUSB Library/src/tusb.h:146:53: error: static assertion failed: CFG_TUSB_RHPORT0_MODE/CFG_TUSB_RHPORT1_MODE must be defined
  146 |   #define _tusb_init_arg0()        TU_VERIFY_STATIC(false, "CFG_TUSB_RHPORT0_MODE/CFG_TUSB_RHPORT1_MODE must be defined")
      |                                                     ^~~~~
.pio/libdeps/pico/Adafruit TinyUSB Library/src/common/tusb_compiler.h:39:31: note: in expansion of macro '_tusb_init_arg0'
   39 | #define TU_STRCAT(a, b)       a##b                ///< concat without expand
      |                               ^
.pio/libdeps/pico/Adafruit TinyUSB Library/src/common/tusb_compiler.h:42:31: note: in expansion of macro 'TU_STRCAT'
   42 | #define TU_XSTRCAT(a, b)      TU_STRCAT(a, b)     ///< expand then concat
      |                               ^~~~~~~~~
.pio/libdeps/pico/Adafruit TinyUSB Library/src/common/tusb_compiler.h:127:43: note: in expansion of macro 'TU_XSTRCAT'
  127 | #define TU_FUNC_OPTIONAL_ARG(func, ...)   TU_XSTRCAT(func##_arg, TU_ARGS_NUM(__VA_ARGS__))(__VA_ARGS__)
      |                                           ^~~~~~~~~~
.pio/libdeps/pico/Adafruit TinyUSB Library/src/tusb.h:151:46: note: in expansion of macro 'TU_FUNC_OPTIONAL_ARG'
  151 | #define tusb_init(...)                       TU_FUNC_OPTIONAL_ARG(_tusb_init, __VA_ARGS__)
      |                                              ^~~~~~~~~~~~~~~~~~~~
/home/frohro/.platformio/packages/framework-arduinopico/libraries/FreeRTOS/src/variantHooks.cpp:501:5: note: in expansion of macro 'tusb_init'
  501 |     tusb_init();
      |     ^~~~~~~~~
In function 'DecodePulseInfo',
    inlined from 'DecodeICS' at .pio/libdeps/pico/libhelix/src/libhelix-aac/noiseless.c:422:3,
    inlined from 'raac_DecodeNoiselessData' at .pio/libdeps/pico/libhelix/src/libhelix-aac/noiseless.c:468:2:
.pio/libdeps/pico/libhelix/src/libhelix-aac/noiseless.c:257:31: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  257 |                 pi->offset[i] = GetBits(bsi, 5);
      |                 ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
In file included from .pio/libdeps/pico/libhelix/src/libhelix-aac/noiseless.c:47:
.pio/libdeps/pico/libhelix/src/libhelix-aac/coder.h: In function 'raac_DecodeNoiselessData':
.pio/libdeps/pico/libhelix/src/libhelix-aac/coder.h:167:19: note: at offset 4 into destination object 'offset' of size 4
  167 |     unsigned char offset[MAX_PULSES];
      |                   ^~~~~~
In function 'DecodePulseInfo',
    inlined from 'DecodeICS' at .pio/libdeps/pico/libhelix/src/libhelix-aac/noiseless.c:422:3,
    inlined from 'raac_DecodeNoiselessData' at .pio/libdeps/pico/libhelix/src/libhelix-aac/noiseless.c:468:2:
.pio/libdeps/pico/libhelix/src/libhelix-aac/noiseless.c:258:28: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  258 |                 pi->amp[i] = GetBits(bsi, 4);
      |                 ~~~~~~~~~~~^~~~~~~~~~~~~~~~~
.pio/libdeps/pico/libhelix/src/libhelix-aac/coder.h: In function 'raac_DecodeNoiselessData':
.pio/libdeps/pico/libhelix/src/libhelix-aac/coder.h:168:19: note: at offset 4 into destination object 'amp' of size 4
  168 |     unsigned char amp[MAX_PULSES];
      |                   ^~~
*** [.pio/build/pico/libbbe/FreeRTOS/variantHooks.cpp.o] Error 1
```  
Maybe it is related to my platformio.ini.  Here is what that looks like.
```
[env:pico]
platform = raspberrypi
board = pico
framework = arduino
board_build.core = earlephilhower
build_flags = -DUSE_TINYUSB
              -DPICO_DEFAULT_CLK_SYS_FREQ_HZ=156300000  ; Set clock speed to 156.3 MHz
            ;   -DCORE_DEBUG_LEVEL=2 ; stands for Warning\
            -DCORE_DEBUG_LEVEL=5 -Wno-unused-variable -Wno-unused-but-set-variable -Wno-unused-function -Wno-format-extra-args   
lib_deps =  https://github.com/pschatzmann/Adafruit_TinyUSB_Arduino.git#Audio
            https://github.com/pschatzmann/arduino-audio-tools.git
            https://github.com/pschatzmann/arduino-libhelix.git 
            https://github.com/greiman/SdFat.git
```
Maybe my problem is related to how I am testing with Audacity (which seems to have been recently updated to 3.6.1).  Here is the code I am using now if I go back to my original code without AudioTools.
```
#include <cmath>
#include "Adafruit_TinyUSB.h"

const float sampleRate = 8000;
const float frequencyLeft = 2000;
const float frequencyRight = 2000;
const float amplitude = 30000.0;
const float twoPi = 2.0 * M_PI;

Adafruit_USBD_Audio usb;

size_t readCB(uint8_t* data, size_t len, Adafruit_USBD_Audio& ref) {
  int16_t* data16 = (int16_t*)data;
  Serial.println(len);
  size_t samples = len / sizeof(int16_t);
  size_t result = 0;
  static float phaseLeft = 0.0;
  static float phaseRight = 0.0;
  float phaseIncrementLeft = twoPi * frequencyLeft / sampleRate;
  float phaseIncrementRight = twoPi * frequencyRight / sampleRate;

  for (size_t j = 0; j < samples; j += 2) {
    data16[j] = amplitude * sin(phaseLeft); // Left channel
    data16[j + 1] = amplitude * sin(phaseRight); // Right channel
    phaseLeft += phaseIncrementLeft;
    phaseRight += phaseIncrementRight;
    if (phaseLeft >= twoPi) phaseLeft -= twoPi;
    if (phaseRight >= twoPi) phaseRight -= twoPi;
    result += sizeof(int16_t) * 2;
  }
  return result;
}

void setup() {
  Serial.begin(115200);

  // Start USB device as Microphone
  usb.setReadCallback(readCB);
  usb.begin(sampleRate, 2, 16); 

  if (TinyUSBDevice.mounted()) {
    TinyUSBDevice.detach();
    delay(10);
    TinyUSBDevice.attach();
  }
}

void loop() {
  // optionally use LED to display status
  usb.updateLED();
}
```
I set the sample rate and frame size by right-clicking in the area just to the left of the audio sample display.  This time to 8 kHz and 16 bits.  In the preferences->Recording, I make sure the "Detect dropouts" is selected.  There may be some issues with Audacity, since it crashes on me sometimes.  I have tried this on two Ubuntu machines, one 24.10 and one newly installed with 24.04.

**Comment by frohro** [comment]
In case it helps, I zipped my PlatformIO project up and put it [here](https://fweb.wallawalla.edu/~frohro/Test_USB_Audio.zip).

**Comment by frohro** [comment]
I am testing on the Ubuntu 24.04 machine recently installed now, building, etc. on it.  The results are the same.  With the Audacity preference->Recoroding-Detect Dropout on here is what it looks like for the first few seconds after plugging in the Pico.  After this, it continues detecting dropouts.  Zoomed out: 
![image](https://github.com/user-attachments/assets/46473090-6110-47d8-bf97-2da953d5ec9f)
and Zoomed in where its behavior changes from good to bad.  
![image](https://github.com/user-attachments/assets/816228af-bb8b-4511-b20e-eda60fc979cc)
You will have to download and unzip it to listen as, Github doesn't seem to allow .wav or any other audio type of file.
[Pico_sound.zip](https://github.com/user-attachments/files/18293994/Pico_sound.zip)

**Comment by pschatzmann** [comment]
Strange: did you set the Project Rate in Audacity at the bottom to the same rate as in the Arduino Sketch ?
Your print screen shows that you have it at 8000 hz instead of the 48000 that you used initially for the tests.

I was testing with Audacity 2.4.2 which comes with Mint.
Recoroding-Detect Dropout was on by default

**Reply by frohro** [reply]
I went to testing at 8 kHz to try and figure out the problem.   I'm still not sure what is wrong with my code, but yours works.  The purpose of my code was to make sure things were working enough to try it out on my PCM1808 ADC hooked up to a Pico.  I'm still testing things a bit with your code.   I did try to go to 96 kHz, but that did not work.  I would like 96 kHz, but 48 kHz will do.

**Comment by frohro** [comment]
@pschatzmann Okay, so I got it to work okay with AudioTools after messing with the platformio.ini.  Your code from AudioTools works well, but mine above has some problem.  Thanks for providing me the example in AudioTools.

In case it is useful to someone trying to get Audio_Tools going with the Pico as a USB sound card, here is my platformio.ini.

```
[env:pico]
platform = raspberrypi
board = pico
framework = arduino
board_build.core = earlephilhower
build_flags = 
    -DUSE_TINYUSB
    -DPICO_DEFAULT_CLK_SYS_FREQ_HZ=156300000  ; Set clock speed to 156.3 MHz
    -DCORE_DEBUG_LEVEL=5 
    -Wno-unused-variable 

    -Wno-unused-function 
    -Wno-format-extra-args
    -DCFG_TUSB_RHPORT0_MODE=OPT_MODE_DEVICE  ; Set USB port 0 as device mode
    -DCFG_TUSB_OS=OPT_OS_PICO               ; Set OS as Pico
    -DCFG_TUSB_DEBUG=0                       ; Debug level
lib_deps =  
    https://github.com/pschatzmann/Adafruit_TinyUSB_Arduino.git#Audio
    https://github.com/pschatzmann/arduino-audio-tools.git
    https://github.com/pschatzmann/arduino-libhelix.git 
    https://github.com/greiman/SdFat.git
```

**Comment by frohro** [comment]
Okay, I found the problem with my code by using a different generator made for finding an off by one error.  It is here:
```
#include "Adafruit_TinyUSB.h"

enum WaveState { TOP, MIDDLE_DOWN, BOTTOM, MIDDLE_UP };
static WaveState state = TOP;
const int16_t TOP_VALUE = 32000;
const int16_t MIDDLE_VALUE = 0;
const int16_t BOTTOM_VALUE = -32000;

int16_t triangleWave = 0;
Adafruit_USBD_Audio usb;

size_t readCB(uint8_t* data, size_t len, Adafruit_USBD_Audio& ref) {
    int16_t* data16 = (int16_t*)data;
    size_t samples = len / sizeof(int16_t);
    size_t result = 0;
    
    for (size_t j = 0; j < samples; j += 2) {
        switch(state) {
            case TOP:
                triangleWave = TOP_VALUE;
                state = MIDDLE_DOWN;
                break;
            case MIDDLE_DOWN:
                triangleWave = MIDDLE_VALUE;
                state = BOTTOM;
                break;
            case BOTTOM:
                triangleWave = BOTTOM_VALUE;
                state = MIDDLE_UP;
                break;
            case MIDDLE_UP:
                triangleWave = MIDDLE_VALUE;
                state = TOP;
                break;
        }
        data16[j] = triangleWave;      // Left channel
        data16[j + 1] = triangleWave;  // Right channel
        result += sizeof(int16_t) * 2;
    }
    return result;
}

void setup() {
    Serial.begin(115200);

    // Start USB device as Microphone
    usb.setReadCallback(readCB);
    usb.begin(44100, 2, 16);

    if (TinyUSBDevice.mounted()) {
        TinyUSBDevice.detach();
        delay(10);
        TinyUSBDevice.attach();
    }
}

void loop() {
    // Update LED to display status
    usb.updateLED();
}
```
The problem was the line:
```
size_t samples = len / sizeof(int16_t)-1;
```
I needed the -1.  @pschatzmann I suspect your original code with the random numbers needs this too, so if someone like me uses it it will work.  The example is in your blog [here](https://www.pschatzmann.ch/home/2024/10/13/tinyusb-audio-on-an-rp2040-in-arduino/).  Perhaps a more accurate way to fix it is to say the for loop goes to samples -1 or maybe in the variable len.  In any case it works perfectly with the -1, but there is one point missing on right and left without it.  It seems like it might be in len.

Here is what it looks like if I make len two less.  USB buffer frames are usually 4 bytes.  When I print len, it is 194 bytes, and that is not evenly divisible by 4.  Here is what my graph looks like with a 1000 Hz sine wave, sampling at 48 kHz with the samples -1.
![image](https://github.com/user-attachments/assets/68660065-0b01-4762-be2d-3a1aa532ef18)

**Comment by pschatzmann** [comment]
Hmm, that's strange: I would not see why size_t samples = len / sizeof(int16_t) should be wrong.
Maybe an assert(len == result) might help to check if we run into some unexpected cases

**Reply by frohro** [reply]
Yes, I agree.  That is correct.  The off by one error is probably back upstream in len or something.
